cwltool/provenance.py

```diff
@@ -0,0 +1,315 @@
+from __future__ import absolute_import
+import io
+import json
+import os
+import os.path
+import posixpath
+import shutil
+import tempfile
+import logging
+import hashlib
+from shutil import copyfile
+import io
+from pathlib2 import Path
+from networkx.drawing.nx_agraph import graphviz_layout
+from networkx.drawing.nx_agraph import write_dot
+from networkx.drawing.nx_pydot import write_dot
+import prov.graph as graph
+import uuid
+import graphviz
+import networkx as nx
+import ruamel.yaml as yaml
+import warnings
+from typing import Any, Dict
+from subprocess import check_call
+warnings.simplefilter('ignore', yaml.error.UnsafeLoaderWarning)
+relativised_input_object={}  # type: Dict[str, Any]
+_logger = logging.getLogger("cwltool")
+
+# RO folders
+METADATA = "metadata"
+DATA = "data"
+OUTPUT= "output"
+WORKFLOW = "workflow"
+SNAPSHOT = "snapshot"
+# sub-folders
+MAIN = os.path.join(WORKFLOW, "main")
+PROVENANCE = os.path.join(METADATA, "provenance")
+
+class ProvenanceException(BaseException):
+    pass
+
+
+# sha1, compatible with the File type's "checksum" field
+# e.g. "checksum" = "sha1$47a013e660d408619d894b20806b1d5086aab03b"
+# See ./cwltool/schemas/v1.0/Process.yml
+hashmethod = hashlib.sha1
+
+class RO():
```

Comment:
I think it is better to keep longer names `ResearchObject` or `research_obj` instead of `RO()` and `ro` everywhere outside the module for easier code understanding.

Reply:
Thanks @anton-khodak for the feedback.  Definitely makes sense. I'd make the required changes.